src/scores/processing/block_bootstrap.py

```diff
@@ -0,0 +1,410 @@
+"""
+Functions for performing block bootstrapping of arrays. This is inspired from
+https://github.com/dougiesquire/xbootstrap with modifications to make the functions more
```

Comment:
I just noticed this, and the implementation is similar enough to https://github.com/dougiesquire/xbootstrap/blob/main/xbootstrap/core.py
that if we want to use it "as-is" we should probably include their license/permission notice.

Alternatively, we should consider re-implementing using a unique solution but using `xbootstrap` as a test case to check if the results are the same. (This obviously would take more time, but is definitely do-able in a more efficient way in my opinion.)

This is one for @tennlee to make the call on.

Reply:
I'll have to really double check this, but my understanding is that the original MIT license and the Apache 2.0 license are compatible. If so there should be a viable way forward here, we'll just need to make sure of how to attribute, reference and include license statements correctly.

Reply:
@reza-armuei I'm going to mark this as "draft" for now, just as a guard.

@tennlee I haven't gone through everything, but my 2c is - it seems there's a few different coding paradigms used in the code, `itertools`, list comprehensions, basic for loops, and `numpy`, rather arbitrarily. A lot of this can just be purely `numpy`, with some exceptions for basic operations - (this would improve clarity and as a bonus speed). The structure of the code seems overly complex for what I believe it's trying to do, and there are a lot of implicit assumptions. The type safety and mutability can also be improved. Having said that it may still be functionally correct in its implementation.

We can discuss more next week.

Reply:
@nikeethr Thank you so much for your time to review and for your comments. I agree with you that the current structure is complex and it can be improved. I have started to apply your comments (it may take time as I am working on a few other tasks too). 
@nikeethr and @tennlee, I am open to any suggestions to improve it.

Reply:
I'll just share this here from summarizing my notes of the algorithm for reference. It'd be easier if we go through a discussion over a call.

---

The way I understand the algorithm (feel free to correct me), is that it is trying to sample blocks (with replacement) essentially to re-arrange/randomize the domain; without breaking any dependence relationship. This can be summarized as:
```
block_sample = {
    (x_1, x_2, ..., x_n)
    such that |x_i - x_i_hat| <= b_i,
    for all x_i, and fixed x_i_hat in each dimension i
}
where,
    i = 1,...,n
    n = number axes
    b_i = dependence distance for axis i (or block size)
    x_i = indices in dimension i
    x_i_hat = index of i'th axis in random sample
```

If we want to add cyclic operations then basically the norm/distance `||` operator will have to account for the space being cyclic (instead of linearly ordered). If there are partial blocks due to block size not being a multiple of domain size - then we either: trim, expand, or provide partial block sampling (not recommended as it will sever dependence).

If this is accurate then here are a following things to consider:
- `numpy.random.default_rng() -> rng.integers()` is actually very performant, and it should be used to sample in bulk.
- It is sufficient to reproduce the algorithm by generating a random point cloud (which is a one-liner), and then expand each point into a block.
- This will result in a array of extents for each axis, which I believe is a perfectly reasonable way of indexing arrays in `xarray/numpy` format.
- If we want sampling to be cyclic, we apply the modulo operator when expanding (suspected partial) blocks. Or a faster (but more memory consuming) way is to just pad the axes.
- If we don't want them to be cyclic, we just restrict the sample domain.
- In order to preserve the "nested" sampling, it is sufficient to preserve
    1. the order (with no jumps) of the dimensions for every input array
    2. for a lower dimensionality we simply ignore the dimensions not present and broadcast to confirm to the other arrays.
- Use `dask` & `apply_ufunc` _optionally (based on user preference/compatibility)_ for the outer repetitions i.e. `num_iterations`

The above seems like a lot of considerations, but honestly the resulting code imo would be quite a bit simpler.